streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java

```diff
@@ -459,12 +461,23 @@ private List<StreamsMetadata> rebuildMetadataForSingleTopology(final Map<HostInf
         return rebuiltMetadata;
     }
 
+    private final Function<Optional<Set<Integer>>, Integer> getPartition = maybeMulticastPartitions -> {
+        if (!maybeMulticastPartitions.isPresent()) {
+            return null;
+        }
+        if (maybeMulticastPartitions.get().size() != 1) {
+            throw new IllegalArgumentException("The partitions returned by StreamPartitioner#partitions method when used for fetching KeyQueryMetadata for key should be a singleton set");
```

Comment:
Thanks for the suggestion. Hmm I had assumed all along that even IQ would be based off of a single partition. Infact we had discussed it during the KIP design and this was a suggestion from Matthias:

```
Btw: The `StreamPartitioner` interface is also used for IQ. For both IQ
and FK-join, it seem ok to just add a runtime check that the returned
list is a singleton (in case we don't add a new class)?
```

But I can see a mismatch in the partitioners used. IIUC, the `StreamPartitioner` used here is to be used for finding the partitioner for the key and I am trying to enforce a single partition check here. But, I don't think I added any such conditions at the time of writing the keys. So, that's something which clearly can causes confusion and even yield wrong results and looks like that's what you are also pointing to.  So, would need to enhance the `KeyQueryMetadata` class. And yeah, deprecating `partition()` method can be avoided at this point.

On a different note, do you think a similar condition can happen even for FK joins?

Reply:
> do you think a similar condition can happen even for FK joins?

Good question -- I had the same thought when I first thought of this but shrugged it off for some reason, now though I think it's worth considering -- what was the rationale behind enforcing a singleton partition for FKJs?

Reply:
i went through the discussion thread again. I think it's just an implicit assumption (mainly by me) that both FK join and IQv2 would work correctly only with a single partitioner. There wasn't much discussion on this though. I will need to check the code for these 2 again to find if that's the case or not.

Reply:
Is it a question of whether it will _only_ work correctly with single partitions, or that we can enforce this without any problems? I guess it seems like the whole point of this KIP is to give users control over whether to send records to other/more partitions, so as long as it doesn't break anything then we should allow it -- of course we still need to figure out whether it does break anything or not

@bellemare or @vvcephei do you know/have any thoughts on this?

Reply:
yeah true. I had assumed that they work only with single partition. It was also driven by the fact that didn't want to touch these paths straight away but when I think about it, maybe it's confusing. Probably Adam/John's inputs might also be helpful here.

Reply:
@ableegoldman , I took a look at this again today and here's what I think:

1) For IQv2, it definitely makes sense to expand the partition to partitions. Given that it is mostly to query data, users can pass in multiple partitions and would definitely want to get them back. There's one nit that I wanted to do point out here, which somehow is also related to multiple partitions. `StreamPartitioner#partition` method states that if it returns a null, then default partition should be used. However, that doesn't somehow get handled in the 2 `StreamsMetadatState#getKeyQueryMetadataForKey` methods. Wrote a simple test case to simulate this and it threw an NPE as expected:

```
    @Test
    public void shouldGetInstanceWithKeyAndCustomPartitionerUsingDefaultPartitioner() {
        final TopicPartition tp4 = new TopicPartition("topic-three", 1);
        hostToActivePartitions.put(hostTwo, mkSet(topic2P0, tp4));
       final StreamPartitioner<String, Object> partitioner1 = (topic, key, value, numPartitions) -> null;

        metadataState.onChange(hostToActivePartitions, hostToStandbyPartitions,
                cluster.withPartitions(Collections.singletonMap(tp4, new PartitionInfo("topic-three", 1, null, null, null))));

        final KeyQueryMetadata expected = new KeyQueryMetadata(hostTwo, Collections.emptySet(), 1);

        final KeyQueryMetadata actual = metadataState.getKeyQueryMetadataForKey("table-three",
                "the-key",
                partitioner1);
        assertEquals(expected, actual);
        assertEquals(1, actual.partition());
    }
```

which throws this:

```
java.lang.NullPointerException
        at org.apache.kafka.streams.processor.internals.StreamsMetadataState.getKeyQueryMetadataForKey(StreamsMetadataState.java:486)
        at org.apache.kafka.streams.processor.internals.StreamsMetadataState.getKeyQueryMetadataForKey(StreamsMetadataState.java:276)
        at org.apache.kafka.streams.processor.internals.StreamsMetadataStateTest.shouldGetInstanceWithKeyAndCustomPartitionerUsingDefaultPartitioner(StreamsMetadataS
```

Is there an implicit assumption that `partition` method won't return null? If not what should be the behaviour in this case? Based on that I would expand it to the `partitions` method as well.

2) I went through the FK join usecase again and I think we should not restrict it to single partition here again. Give that the FK would be used for partitioner for co-partitioning purposes, it's very well possible that the FK might be across multiple partitions and the composite key should also be produced across those partitions for effective joining. 


I would update the KIP once we hear back from John/Adam. Thanks!

Reply:
Hi all, on question (2), I'm not sure that it'll be straightforward to establish the semantics of FK Join on multicasted tables. Personally, I'd be more comfortable just keeping the restriction that we see in this PR and following up with a separate KIP/PR to lift that restriction.

I don't think that restriction will directly harm the benefit we're trying to achieve with this KIP, and it'll be relatively complicated to establish what a multicasted FK join means and whether it is correct.

Reply:
Thanks @vvcephei . Point taken!